py3status/modules/updates.py

```diff
@@ -0,0 +1,199 @@
+        if output is None:
+            return {self.name: output}
+        return {self.name: self.count_updates(output)}
+
+
+class Apt(Update):
+    def count_updates(self, output):
+        return len(output.splitlines()[1:])
+
+
+class Apk(Update):
+    def count_updates(self, output):
+        return len(output.splitlines()[1:])
+
+
+class Eopkg(Update):
+    def get_output(self):
+        output = self.parent.py3.command_output(self.command)
+        if "No packages to upgrade." in output:
+            return ''
+        return output
+
+
+class Zypper(Update):
+    def count_updates(self, output):
+        return len([x for x in output.splitlines() if x][4:])
+
+
+class Py3status:
+    """
+    """
+
+    # available configuration parameters
+    cache_timeout = 600
+    format = None
+    thresholds = [(0, 'darkgray'), (10, 'degraded'),
+                  (20, 'orange'), (30, 'bad')]
+
+    class Meta:
+        deprecated = {
+            "rename_placeholder": [
+                {
+                    "placeholder": "aur",
+                    "new": "auracle",
+                    "format_strings": ["format"],
+                }
+            ]
+        }
+
+    def post_config_hook(self):
+        managers = [
+            ("pacman", "checkupdates"),
+            ("auracle", "auracle sync"),
+            ("yay", "yay --query --upgrades --aur"),
+            ("apk", "apk version -l '<'"),
+            ("apt", "apt list --upgradeable"),
+            ("eopkg", "eopkg list-upgrades"),
+            ("pkg", "pkg upgrade --dry-run --quiet"),
+            ("xbps", "xbps-install --update --dry-run"),
+            ("zypper", "zypper list-updates"),
+        ]
+
+        managed = getattr(self, 'managers', [])
+        if managed:
+            if isinstance(managed[0], tuple):
+                managers = managed
+            else:
+                new_managers = []
+                for name in managed:
+                    for manager in managers:
+                        if manager[0] == name:
+                            new_managers.append(manager)
+                            break
+                managers = new_managers
+        else:
+            placeholders = self.py3.get_placeholders_list(self.format or '')
+            managed = [x for x in placeholders if x != 'update']
+
+        names = []
+        self.backends = []
+        for name, command in managers:
+            if self.py3.check_commands(command.split()[0]):
+                names.append(name)
+                try:
+                    backend = globals()[name.capitalize()]
+                except KeyError:
+                    backend = Update
+                self.backends.append(backend(self, name, command))
+
+        if not self.format:
+            auto = "[\?not_zero {Name} [\?color={name} {{{name}}}]]"
+            self.format = "[{}|\?show No Updates]".format("[\?soft  ]".join(
+                auto.format(Name=x.capitalize(), name=x) for x in names
+            ))
```

Comment:
why is this logic here and not in the beginning where we declare default values for `format`?

Reply:
Shouldn't the default value of `format` be something like `[{format_manager}|\?show No Updates]`? And `auto` in turn to be `format_manager` option?

Reply:
It's not a list of composite. It's similar to `sysdata` and everything is in one `format`. It is probably possible to make generic `format_update` and...
* `format_pacman`
* `format_eopkg`
* `format_apt`
* `format_apk`
* `format_zypper`
* `format_xbps`
et cetera, et cetera.

In each `format_*`, they would have their own placeholders and are 100% different from each others. Comparing this to `mega` :wink:, this would be `{sync}` placeholder on your `Format placeholders`... where it would print number of `sync` lines in the output (i.e mostly 1).

Reply:
`{format_sync}` is what I thought as well 🙂 But I think having multiple specific formats is not needed, just `format_manager` in which people can use `{name}` and `{count}`, so that instead of (just for the sake of example) default `format_manager = '{name} {count}'` they can wrap `{count}` in some color — it should apply to all managers.

Reply:
>having multiple specific formats is not needed

It is needed because every update manager will be different from each others. That's why it is in classes so it could be extended to support this if we want to.... That would be the crazy part.


Single `command | shell | shell shell` wouldn't fly here unless we make our own class... then in that case, we should just use `command` + python code instead.

I understand what you're saying, but we can't guarantee that. I think `apt` have its own `install`, `remove`, `upgrade`, `keep` counts so technically, all placeholders here should have `_update` suffix if we were to expand this some more.
* `apt_install`   # install implies additional new packages
* `apt_upgrade`
* `apt_remove`
* `pacman_update`
* `auracle_update`
* `zypper_install`
*  Bunch of `format_*` for per-package loops.

The best part about `format_*` is that you can add random bar/block icon and watch it grow... or in `arch_updates`, you set up colorized bars (citing my own `arch_updates` PR).

Reply:
IMHO if we are making a generic `updates` module, we need to align all different package managers to one standard, i.e. not support special cases of some package managers. In the case of `apt`, I would say showing `_remove` and `_install` is out of scope and should not be supported, it doesn't matter if by upgrading `3` packages your will additionally install `2` dependencies,  this module is about showing how many packages need to be updated.

Reply:
And if someone really really wants to additionally show the number of install packages, they can add this as a new manager 😉

Reply:
This might not work because we can't all query same manager... And lot of time, doing `format_*` requires fixing lines, slicing lines, etc. Anyhow, I think I fixed this to do what you want... with `managers` + `format`. Care to give it a go?

Reply:
I still don't understand why `format` is declared as default value of "None", where it's clearly `[{format_managers}|\?show No Updates]`...

I expected to see something like this in the beginning of the file:

```
class Py3status:
    format_manager = "[\?not_zero {manager} [\?color={count} {{{count}}}]]"
    format = "[{format_managers}|\?show No Updates]"
```